# HG changeset patch
# User Pantxo Diribarne <pantxo.diribarne@gmail.com>
# Date 1522097664 -7200
#      Mon Mar 26 22:54:24 2018 +0200
# Branch stable
# Node ID 64c6e8bfc9e5fddbaeed9c0c7ec3992c26f7fd7c
# Parent  7f1d1a1ca22ef91bf18654b4b9dbb17bd3ab21d8
Fix Octave not loading figures saved in older versions (bug #53468)

* struct2hdl.m: Silence deprecation warnings
  (setprops): add missing properties to labels before trying to set them
  (addmissingprops): make use of isprop to include hidden properties in the
  search

diff --git a/scripts/plot/util/struct2hdl.m b/scripts/plot/util/struct2hdl.m
--- a/scripts/plot/util/struct2hdl.m
+++ b/scripts/plot/util/struct2hdl.m
@@ -110,16 +110,19 @@ function [h, pout] = struct2hdl (s, p=[]
     else
       names{n_pos(1)} = "position";
       names{n_pos(2)} = "outerposition";
     endif
   endif
   ## Reorder the properties with the mode properties coming last
   s.properties = orderfields (s.properties, names);
 
+  ## Silence deprecation warnings
+  warning ("off", "Octave:deprecated-property", "local");
+  
   ## create object
   if (strcmp (s.type, "root"))
     h = 0;
     s.properties = rmfield (s.properties, ...
                               {"callbackobject", "commandwindowsize", ...
                                "screendepth", "screenpixelsperinch", ...
                                "screensize"});
   elseif (strcmp (s.type, "figure"))
@@ -590,18 +593,19 @@ function setprops (s, h, p, hilev)
       nf = length (idx);
       fields = fields(idx);
       vals = vals(idx);
       while (nf)
         field = fields{nf};
         idx = find (hdls == vals{nf});
         spec = specs(idx);
         if (isprop (h, field))
-           h2 = get (h , field);
-           set (h2, spec.properties);
+          h2 = get (h , field);
+          addmissingprops (h2, spec.properties);
+          set (h2, spec.properties);
         endif
         nf -= 1;
       endwhile
 
       ## If hggroup children were created by high level functions,
       ## copy only useful properties.
       if (hilev)
         if (strcmp (s.type, "hggroup"))
@@ -643,22 +647,19 @@ function out = valcomp (x, hdls)
   endif
 
 endfunction
 
 function addmissingprops (h, props)
 
   hid = {"__autopos_tag__", "looseinset"};
   oldfields = fieldnames (props);
-  curfields = fieldnames (get (h));
-  missing = ! ismember (oldfields, curfields);
-  idx = find (missing);
-  for ii = 1:length (idx)
-    prop = oldfields{idx(ii)};
-    if (! any (strcmp (prop, hid)))
+  for ii = 1:numel (oldfields)
+    prop = oldfields{ii};
+    if (! isprop (h, prop) && ! any (strcmp (prop, hid)))
       addproperty (prop, h, "any");
     endif
   endfor
 
 endfunction
 
 
 ## FIXME: Need validation tests
